src/vivarium/framework/randomness/stream.py

```diff
@@ -25,19 +24,26 @@
 
 """
 
+from __future__ import annotations
+
 import hashlib
-from typing import Any, Callable, List, Optional, Tuple, Union
+from typing import Any, Callable, TypeVar
 
 import numpy as np
+import numpy.typing as npt
 import pandas as pd
 from scipy import stats
 
 from vivarium.framework.randomness.exceptions import RandomnessError
 from vivarium.framework.randomness.index_map import IndexMap
 from vivarium.framework.utilities import rate_to_probability
+from vivarium.types import ClockTime, NumericArray
 
 RESIDUAL_CHOICE = object()
 
+# TODO: Parameterizing pandas objects fails below python 3.12
+P = TypeVar("P", pd.DataFrame, pd.Series, pd.Index)  # type: ignore [type-arg]
```

Comment:
Can you give this TypeVar a more informative name? Maybe something like `PandasObject`?